### constitutive vs specific of prna from five type cell line

###################################################### to get cell line specific barplot 
##pca data are counted by feature count use annotation of merge (refeseq_curated_nm.gtf and ps/pas.gtf)
##refseq_curated_nm.gtf are generated by ucsc table browser group:'genes and genes prediction',track:'ncbi refseq'
##followed grep 'gene_id\s"NM_' refseq_curated.gtf > refseq_curated_nm.gtf
####after  new annotation output count data , transform sample encode id to cell line id
library(dplyr)
library(tibble)
library(reshape2)
ps = read.csv("H:/prna_transcript_analysis/PCA/pas_genes_count.csv",header = TRUE,row.names = 1)#from pca/featurecount/p(a)s/ directory
pas = read.csv("H:/prna_transcript_analysis/PCA/ps_genes_count.csv",header = TRUE,row.names = 1)
rownames(ps) = gsub('^','ps_',rownames(ps))
rownames(pas) = gsub('^','pas_',rownames(pas))
output = rbind(ps,pas)
ps_length = read.csv('H:/prna_transcript_analysis/PCA/ps_genelength',sep = '\t',skip = 1)
pas_length = read.csv('H:/prna_transcript_analysis/PCA/pas_genelength',sep = '\t',skip = 1)
tpm <- function(raw_counts, gene_lengths) {
  x <- raw_counts*1e3 / gene_lengths
  return(t(t(x)*1e6 / colSums(x)))
  
}
length = c(ps_length$Length,pas_length$Length)
tpm_output = tpm(output,length)



colddata = read.csv("H:/prna_transcript_analysis/epd_promoter/encode_data_pca/coldata.csv")
colddata$X = gsub("\\..*","",colddata$X)
mix = as.data.frame(t(merge(t(tpm_output),colddata, by.x = 'row.names',by.y = 'X')))
colnames(mix) = mix[rownames(mix) %in% c('condition'),]
mix = mix[!rownames(mix) %in% c("Row.names","type","cell","condition"),]


mix_ps = mix[grep('ps',rownames(mix)),]
mix_pas = mix[grep('pas',rownames(mix)),]
####cell line specific expressed prna bar plot
mix_ps =  mix_ps[,order(names(mix_ps))]
mix_pas =  mix_pas[,order(names(mix_pas))]
stat1 = apply(apply(mix_ps,2,function(x) as.numeric(x)),2,function(x) table(x>=5))##here set cpm cutoff
stat2 = apply(apply(mix_pas,2,function(x) as.numeric(x)),2,function(x) table(x>=5))

### replace stat1/stat2 to plot bar scheme
melt_stat1 = melt(stat1) %>% filter(Var1 == 'TRUE')
melt_stat1$group = gsub("\\..*","",melt_stat1$Var2)
library(ggplot2)
plot_melt_stat1 = melt_stat1 %>% arrange(desc(value)) %>% mutate(Var2= factor(Var2, levels=Var2))
mypal <- c('#BFA2DB','#0E5E6F','#DD5353','#FFD4D4','#829460','#F2F7A1','#F2CD5C','#ADDDD0','#D36B00','#319DA0',
          '#7895B2','#E8DFCA','#6FEDD6','#4B5D67','#E3ACF9','#CCD6A6','#90A17D','#57CC99',
          '#A77979','#D6CDA4','#285430','#91D8E4','#557153')

p = ggplot(data=plot_melt_stat1, aes(x=Var2,weight = value,fill=group)) +geom_bar(position='dodge')+ 
  theme(axis.text.x=element_blank())+labs(x = "Samples")+labs(y = "Expressed psRNA Count")+
  theme(panel.border = element_rect(fill=NA,color="black", size=0.5, linetype="solid"),
        axis.text=element_text(size=19),
        axis.title=element_text(size=38,face="bold"))+
  theme(legend.text = element_text(size = 20),legend.title = element_text(size = 20),panel.background = element_blank())+scale_fill_manual(values = mypal)

#1009 595
ggsave('ps_bar.png',p,width = 17,height = 10)


melt_stat2 = melt(stat2) %>% filter(Var1 == 'TRUE')
melt_stat2$group = gsub("\\..*","",melt_stat2$Var2)
library(ggplot2)
plot_melt_stat2 = melt_stat2 %>% arrange(desc(value)) %>% mutate(Var2= factor(Var2, levels=Var2))
mypal <- c('#BFA2DB','#0E5E6F','#DD5353','#FFD4D4','#829460','#F2F7A1','#F2CD5C','#ADDDD0','#D36B00','#319DA0',
                    '#7895B2','#E8DFCA','#6FEDD6','#4B5D67','#E3ACF9','#CCD6A6','#90A17D','#57CC99',
                    '#A77979','#D6CDA4','#285430','#91D8E4','#557153')
                    
p = ggplot(data=plot_melt_stat2, aes(x=Var2,weight = value,fill=group)) +geom_bar(position='dodge')+ 
  theme(axis.text.x=element_blank())+labs(x = "Samples")+labs(y = "Expressed pasRNA Count")+
  theme(panel.border = element_rect(fill=NA,color="black", size=0.5, linetype="solid"),
        axis.text=element_text(size=19),
        axis.title=element_text(size=38,face="bold"))+
  theme(legend.text = element_text(size = 20),legend.title = element_text(size = 20),panel.background = element_blank())+scale_fill_manual(values = mypal)

#1009 595
ggsave('pas_bar.png',p,width = 17,height = 10)














############################################# plot 5 set venn diagram from cpm normalzied data ,
#feature_count annotation = ucsc downloaded hg38 ncbi curated genes and grem NM only + prna   
sense = tpm_output[grep('ps_',rownames(tpm_output)),]
coldata = read.csv("H:/prna_transcript_analysis/epd_promoter/encode_data_pca/sense/single_coldata.csv")
coldata$X = gsub("\\..*","",coldata$X)
#coldata = coldata[grep(pattern = 'iPSC|foreskin_fiber|lymphoblasts', x = coldata$cell),]
sense = sense[,which(colnames(sense) %in% coldata$X)]
coldata = coldata[order(coldata$X),]
sense = as.data.frame(sense[,order(colnames(sense))])
colnames(sense) = coldata$cell

stem_cell = sense[,grep("stem_cell",colnames(sense))]
stem_cell$mean_stem_cell = rowMeans(stem_cell)
stem_cell = filter(stem_cell,mean_stem_cell>=10)

foreskin_fiber = sense[,grep("foreskin_fiber",colnames(sense))]
foreskin_fiber$mean_foreskin_fiber = rowMeans(foreskin_fiber)
foreskin_fiber = filter(foreskin_fiber,mean_foreskin_fiber>=10)

prostate = sense[,c(35,36)]

prostate = filter(prostate,prostate>=10)

breast = sense[,grep("breast",colnames(sense))]
breast$mean_breast = rowMeans(breast)
breast = filter(breast,mean_breast>=10)

lymphoblasts = sense[,grep("lymphoblasts",colnames(sense))]
lymphoblasts$mean_lymphoblasts = rowMeans(lymphoblasts)
lymphoblasts = filter(lymphoblasts,mean_lymphoblasts>=10)

x = list(a = rownames(stem_cell),
         b = rownames(foreskin_fiber),
         c = rownames(prostate),
         d = rownames(lymphoblasts),
         e = rownames(breast))

##plot 5 set venn diagram
library("extrafont")
library(ggVennDiagram)
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}
display_venn(
  x,
  category.names = c("stem_cell" , "foreskin_fiber " , "prostate", "lymphoblasts","breast"),
  # Circles
  lwd = 2,
  lty = 'blank',
  fill = c("#e6c994", "#E69F00", "#d4bbd8", "#086e7d","#51a16a"),
  # Numbers
  cex = 1.7,
  #fontface = "arial",
  fontfamily ="arial",
  # Set names
  cat.cex = 1.9,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.dist = c(0.055, 0.081, 0.1, 0.1,0.1)
)
ggsave(filename = 'venn.png',p,width = 8,height = 8)




#### plot pca for encode total stranded RNA-seq data
rm(list=ls())
library(ggrepel)
library(ggplot2)
library(DESeq2)
library(ggthemes)
library(tidyr)
library(dplyr)
sense = read.csv("H:/prna_transcript_analysis/PCA/pas_output.txt",sep = '\t',row.names = 1)
coldata = read.csv("H:/prna_transcript_analysis/epd_promoter/encode_data_pca/sense/single_coldata.csv")
#coldata = coldata[grep(pattern = 'iPSC|foreskin_fiber|lymphoblasts', x = coldata$cell),]
coldata$condition = gsub("-","_",coldata$condition)
sense = sense[,which(colnames(sense) %in% coldata$X)]
coldata = coldata[order(coldata$X),]
sense = sense[,order(colnames(sense))]
#colnames(sense) = coldata$cell
##delete abnormal data
sense = sense[,-c(2,8,22)]
coldata = coldata[-c(2,8,22),]
dds <- DESeqDataSetFromMatrix(countData = sense,
                              colData = coldata,
                              design = ~ cell)

keep <- rowSums(counts(dds)) >= 5
dds <- dds[keep,]
vsd <- vst(dds, blind=FALSE)
#if error in antisense
#vsd = varianceStabilizingTransformation(dds,blind=FALSE)
#rld <- rlog(dds, blind=FALSE)
pca = plotPCA(vsd, intgroup=c("cell"),returnData = TRUE)
p = ggplot(pca, aes(x=PC1, y=PC2,color=group))+ geom_point(size=2)+theme_bw()+theme(panel.border=element_blank(),panel.grid.major=element_blank(),
                                                                                  panel.grid.minor=element_blank(),
                                                                                  axis.line= element_line(colour = "black"),
                                                                                 axis.text = element_text(size = 20),
                                                                                  legend.title = element_text(size=20),
                                                                                  legend.text = element_text(size=20),
                                                                                  axis.title.x = element_text(size = 20),
                                                                                  axis.title.y = element_text(size = 20),
                                                                                  plot.title = element_text(hjust = 0.4))+
  labs(title = "pasRNA PCA")+ stat_ellipse(level = 0.999)
#+geom_text_repel (aes(PC1,PC2, label=group))  ###add text lable on every dot
###669 473

ggsave('pca.png',p,width = 10,height = 10)




### known lncRNA expression statistics prna
### known lncRNA are download from lncipedia and bedtools intersect with ps.bed
pas_lnc = read.csv("H:/ps_lnc.bed",header = FALSE,sep = '\t')
pas_lnc_expression  = t_merge[which(pas_lnc$V4 %in% rownames(t_merge)),]
write.csv(pas_lnc_expression,"H:/pas_lnc_expression.csv",row.names = FALSE)
singledf = read.csv("H:/pas_lnc_expression.csv")

singledf = singledf[ , order(names(singledf))]
library(dplyr)
stat = apply(singledf,2,function(x) as.numeric(x))
stat = apply(singledf,2,function(x) table(x>=0.5))
library(reshape2)
melt_stat = melt(stat)
melt_stat = melt_stat[grep('TRUE',melt_stat$Var1),]
melt_stat$group = gsub("\\..*","",melt_stat$Var2)
write.csv(melt_stat,"H:/melt_stat.csv")
melt_stat = melt_stat[-grep('IMR.90.1|H1.1',melt_stat$Var2),]


library(ggplot2)
library(cartography)
mypal <- carto.pal(pal1 = "pastel.pal", n1 = 20,middle = TRUE, transparency = FALSE)
ggplot(data=melt_stat, aes(x=Var2,weight = value,fill=group)) +geom_bar(position='dodge')+ theme(axis.text.x=element_blank())+labs(x = "Samples")+labs(y = "Expressed psRNA Count")+
  theme(panel.border = element_rect(fill=NA,color="black", size=0.5, linetype="solid"),axis.text=element_text(size=19),axis.title=element_text(size=18,face="bold"))+
  theme(legend.text = element_text(size = 20),legend.title = element_text(size = 20),panel.background = element_blank())+scale_fill_manual(values = mypal)

#1009 595
